### DISCUSSION TITLE
v0.11.10 Testing checklist

### DISCUSSION THREAD

**Discussion by jorgeepc** [kubeshop/tracetest#2623] [Release testing]
# Testing Tracetest

This is the entire checklist on what we should do to assert that Tracetest is working fine on each version release. On each version release, we can copy the contents of this template checklist and open a Github Discussion to start each test.

On Tracetest, we work in two ways to test the entire system and guarantee that everything is working fine. One is through automatic tests, where we run unit, integration and end-to-end tests on Tracetest CLI, Web UI and Server.

Another source of tests is the manual tests that we execute on each release, simulating users, and checking if everything is ok on these tests.

## Manual Tests

On our manual tests, we aim to do some [sanity checks](https://en.wikipedia.org/wiki/Sanity_check) to see if the main features are working as expected. Usually, we run it on each Tracetest release.

### Tests to validate on Release Candidate

- Test server installation via CLI
  - [x] Docker Compose and no demo API
  - [x] Docker Compose and demo API
  - [x] Kubernetes and no demo API
  - [x] Kubernetes and demo API
  - [x] Verify that Tracetest is not sending analytics if you have the `TRACETEST_DEV` env for docker install
  - [x] Verify that Tracetest is not sending analytics if you have the `TRACETEST_DEV` env for k8s install

- Test DataStore provisioning and basic test
  - [x] Open WebUI and go to `/settings` page. The provisioned Data Store should be selected
  - [x] Create a test on WebUI that calls a demo API (like [Pokeshop](https://docs.tracetest.io/live-examples/pokeshop/overview) or [Open Telemetry Store](https://docs.tracetest.io/live-examples/opentelemetry-store/overview)). This test should fetch traces correctly and run without errors

- Test Tracetest examples
  - [x] [Jaeger example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-jaeger)
  - [x] [Tempo example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-tempo)
  - [x] [New Relic example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-new-relic)
  - [x] [Lightstep example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-lightstep)
  - [x] [OpenSearch example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-opensearch)
  - [x] [Elastic APM example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-elasticapm-with-elastic-agent)
  - [x] [SignalFX example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-signalfx)
  - [x] [Datadog example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-datadog)
  - [x] [Amazon X-Ray example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-amazon-x-ray)
  - [x] [Honeycomb](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-honeycomb)

- Test specific features added/changed on this release:
  - [x] validate Trace Analyzer settings from webUI. Enable/disable. Update minimum score
  - [x] validate a Test run with the Trace Analyzer enabled. Verify the Trace tab
  - [x] validate a Test run with the Trace Analyzer disabled. Verify the Trace tab
    - ux issue: https://github.com/kubeshop/tracetest/issues/2633
  - [ ] validate a Test run (with specs) that passes the minimum score in the Trace Analyzer
     - bug: https://github.com/kubeshop/tracetest/issues/2634
  - [ ] validate a Test run (with specs) that fails the minimum score in the Trace Analyzer
    - bug: https://github.com/kubeshop/tracetest/issues/2636
  - [x] enable/disable ssl cert verification in test trigger https://github.com/kubeshop/tracetest/pull/2521
  - [x] flag and parameter validations for CLI `apply` command https://github.com/kubeshop/tracetest/pull/2600
  - [x] flag and parameter validations for CLI `configure` command https://github.com/kubeshop/tracetest/pull/2600
  - [x] flag and parameter validations for CLI `delete` command https://github.com/kubeshop/tracetest/pull/2600
  - [x] flag and parameter validations for CLI `get` command https://github.com/kubeshop/tracetest/pull/2600
  - [x] flag and parameter validations for CLI `list` command https://github.com/kubeshop/tracetest/pull/2600
  - [ ] flag and parameter validations for CLI `server install` command https://github.com/kubeshop/tracetest/pull/2600
  - [x] validate `All gRPC Spans: Status is Ok` test spec snippet https://github.com/kubeshop/tracetest/pull/2605
  - [x] validate server and cli version match https://github.com/kubeshop/tracetest/pull/2598
  - [x] validate protobuf file path reference on Test yaml from CLI https://github.com/kubeshop/tracetest/pull/2620
  
### Tests to validate on Final Release

- Pipeline
  - [x] Check if our release pipeline on [Release Tracetest](https://github.com/kubeshop/tracetest/actions/workflows/release-version.yml) workflow on Github Actions worked correctly

- Test CLI updates
  - [x] MacOS via homebrew
  - [x] MacOS via curl script
  - [ ] Windows via chocolatey
  - [ ] Windows via manual download

- Docs
  - [ ] Double check [Detailed installation](https://docs.tracetest.io/getting-started/detailed-installation) docs and see if everything is documented correctly

- Test specific features added/changed on this release:
  - [ ] Add list of features here
  
## Automatic Tests

Today Tracetest has 3 main components: a WebUI, a CLI and a Server.

### Web UI

- **Unit tests**: Run by executing `npm test` on `./web` folder
- **End-to-end tests**: Run using [cypress](https://www.cypress.io/) against a temporary Tracetest created on Kubernetes. 

### CLI

- **Unit tests**: Run by executing `make test` on `./cli` folder

### Server

- **Unit tests**: Run by executing `make test` on `./server` folder
- **Integration tests**: Run along with some unit tests running `make test` on `./server` folder, but also done in a matrix test on Github actions, by executing the `test-examples` action.
- **End-to-end tests**: Run using Tracetest to test itself. We deploy two instances of Tracetest and use one to test API calls to another on Github actions, by executing the `trace-testing` action.

**Comment by jorgeepc** [comment]
Bug list
- Minor bug in cli `tracetest configure`: #2632
- `tracetest server install` not using param validations: #2635 2635
- https://github.com/kubeshop/tracetest/issues/2634
- https://github.com/kubeshop/tracetest/issues/2633
- https://github.com/kubeshop/tracetest/issues/2636

**Comment by xoscar** [comment]
The temporary root span bug issue

#2637

**Comment by danielbdias** [comment]
Error on `tracetest test run`: https://github.com/kubeshop/tracetest/issues/2645
